File: src/main/java/info/skyblond/archivedag/service/impl/UserManagementService.java
package info.skyblond.archivedag.service.impl;

import info.skyblond.archivedag.model.DuplicatedEntityException;
import info.skyblond.archivedag.model.EntityNotFoundException;
import info.skyblond.archivedag.model.PermissionDeniedException;
import info.skyblond.archivedag.model.UserDetailModel;
import info.skyblond.archivedag.model.entity.UserEntity;
import info.skyblond.archivedag.model.entity.UserRoleEntity;
import info.skyblond.archivedag.repo.UserRepository;
import info.skyblond.archivedag.repo.UserRoleRepository;
import info.skyblond.archivedag.util.SecurityUtils;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.LinkedList;
import java.util.List;

@Service
public class UserManagementService {
    private final UserRepository userRepository;
    private final UserRoleRepository userRoleRepository;
    private final PasswordEncoder passwordEncoder;
    private final PatternService patternService;

    public UserManagementService(UserRepository userRepository, UserRoleRepository userRoleRepository, PasswordEncoder passwordEncoder, PatternService patternService) {
        this.userRepository = userRepository;
        this.userRoleRepository = userRoleRepository;
        this.passwordEncoder = passwordEncoder;
        this.patternService = patternService;
    }

    public List<String> listUsername(String keyword, Pageable pageable) {
        List<String> result = new LinkedList<>();
        this.userRepository.findAllByUsernameContaining(keyword, pageable)
                .forEach(u -> result.add(u.getUsername()));
        return result;
    }

    public UserDetailModel queryUser(String username) {
        UserEntity entity = this.userRepository.findByUsername(username);
        if (entity == null) {
            return null;
        }
        List<String> roles = new LinkedList<>();

        this.userRoleRepository.findAllByUsername(username)
                .forEach(r -> roles.add(r.getRole()));

        return new UserDetailModel(
                entity.getUsername(),
                entity.getStatus(),
                roles
        );
    }

    public List<String> listUserRoles(String username, Pageable pageable) {
        List<String> roles = new LinkedList<>();
        this.userRoleRepository.findAllByUsername(username, pageable)
                .forEach(r -> roles.add(r.getRole()));
        return roles;
    }

    @Transactional
    public void changePassword(String username, String password) {
        if (this.userRepository.findByUsername(username) == null) {
            throw new EntityNotFoundException("User " + username);
        }
        String encodedPassword = this.passwordEncoder.encode(password);
        this.userRepository.updateUserPassword(username, encodedPassword);
    }

    @Transactional
    public void changeStatus(String username, UserEntity.UserStatus status) {
        if (this.userRepository.findByUsername(username) == null) {
            throw new EntityNotFoundException("User " + username);
        }
        this.userRepository.updateUserStatus(username, status);
    }

    @Transactional
    public void createUser(String username, String password) {
        if (!this.patternService.isValidUsername(username)) {
            throw new IllegalArgumentException("In valid username. The username must meet the regex: " + this.patternService.getUsernameRegex());
        }
        UserEntity entity = new UserEntity(username, this.passwordEncoder.encode(password));
        // Here we should use lock to ensure the user still doesn't exist
        // after this execution, but seems like a lock can only be applied
        // to select statement in JPA, and transaction cannot ensure this
        // unless the isolation is sequence.
        if (this.userRepository.existsByUsername(username)) {
            throw new DuplicatedEntityException("user");
        }
        // The issue for save is: If the username not exists, then it's an insert
        // But if the username exists, it updates.
        this.userRepository.save(entity);
    }

    @Transactional
    public void deleteUser(String username) {
        if (this.userRepository.findByUsername(username) == null) {
            throw new EntityNotFoundException("User " + username);
        }
        // TODO check other resources, like cert
        this.userRoleRepository.deleteAllByUsername(username);
        this.userRepository.deleteByUsername(username);
    }

    @Transactional
    public void addRoleToUser(String username, String role) {
        if (this.userRepository.findByUsername(username) == null) {
            throw new EntityNotFoundException("User " + username);
        }
        UserRoleEntity entity = new UserRoleEntity(username, role);
        if (this.userRoleRepository.exists(Example.of(entity))) {
            throw new DuplicatedEntityException("role for user");
        }
        this.userRoleRepository.save(entity);
    }

    @Transactional
    public void removeRoleFromUser(String username, String role) {
        if (this.userRepository.findByUsername(username) == null) {
            throw new EntityNotFoundException("User " + username);
        }
        if (!this.userRoleRepository.existsByUsernameAndRole(username, role)) {
            throw new EntityNotFoundException("Role " + role + " for user " + username);
        }
        this.userRoleRepository.deleteByUsernameAndRole(username, role);
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arstue/UserManagementService.kt
package info.skyblond.archivedag.arstue

import info.skyblond.archivedag.arstue.entity.UserEntity
import info.skyblond.archivedag.arstue.entity.UserRoleEntity
import info.skyblond.archivedag.arstue.model.UserDetailModel
import info.skyblond.archivedag.arstue.repo.CertRepository
import info.skyblond.archivedag.arstue.repo.UserRepository
import info.skyblond.archivedag.arstue.repo.UserRoleRepository
import info.skyblond.archivedag.arstue.service.PatternService
import info.skyblond.archivedag.commons.DuplicatedEntityException
import info.skyblond.archivedag.commons.EntityNotFoundException
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class UserManagementService(
    private val userRepository: UserRepository,
    private val userRoleRepository: UserRoleRepository,
    private val certRepository: CertRepository,
    private val passwordEncoder: PasswordEncoder,
    private val patternService: PatternService,
    private val fileRecordService: FileRecordService
) {
    fun listUsername(keyword: String, pageable: Pageable): List<String> {
        val result: MutableList<String> = LinkedList()
        userRepository.findAllByUsernameContaining(keyword, pageable)
            .forEach { result.add(it.username) }
        return result
    }

    fun userExists(username: String): Boolean {
        return userRepository.existsByUsername(username)
    }

    fun queryUser(username: String): UserDetailModel {
        val entity = userRepository.findByUsername(username)
            ?: throw EntityNotFoundException("User $username")
        val roles: MutableList<String> = LinkedList()
        userRoleRepository.findAllByUsername(username)
            .forEach { roles.add(it.role) }
        return UserDetailModel(
            entity.username,
            entity.status,
            roles
        )
    }

    fun listUserRoles(username: String, pageable: Pageable): List<String> {
        val roles: MutableList<String> = LinkedList()
        userRoleRepository.findAllByUsername(username, pageable)
            .forEach { roles.add(it.role) }
        return roles
    }

    @Transactional
    fun changePassword(username: String, password: String) {
        if (!userRepository.existsByUsername(username)) {
            throw EntityNotFoundException("User $username")
        }
        val encodedPassword = passwordEncoder.encode(password)
        userRepository.updateUserPassword(username, encodedPassword)
    }

    @Transactional
    fun changeStatus(username: String, status: UserEntity.Status) {
        if (!userRepository.existsByUsername(username)) {
            throw EntityNotFoundException("User $username")
        }
        userRepository.updateUserStatus(username, status)
    }

    @Transactional(isolation = Isolation.SERIALIZABLE)
    fun createUser(username: String, password: String) {
        require(patternService.isValidUsername(username)) { "In valid username. The username must meet the regex: " + patternService.usernameRegex }
        val entity = UserEntity(username, passwordEncoder.encode(password))
        if (userRepository.existsByUsername(username)) {
            throw DuplicatedEntityException("user")
        }
        userRepository.save(entity)
    }

    @Transactional
    fun deleteUser(username: String) {
        if (!userRepository.existsByUsername(username)) {
            throw EntityNotFoundException("User $username")
        }
        // if there are file records, refuse to delete
        if (fileRecordService.listOwnedRecords(username, Pageable.ofSize(10)).isNotEmpty()) {
            throw IllegalStateException("User $username still have file records")
        }
        // delete certs
        certRepository.deleteAllByUsername(username)
        // delete roles
        userRoleRepository.deleteAllByUsername(username)
        // delete user
        userRepository.deleteByUsername(username)
    }

    @Transactional
    fun addRoleToUser(username: String, role: String) {
        if (!userRepository.existsByUsername(username)) {
            throw EntityNotFoundException("User $username")
        }
        val entity = UserRoleEntity(username, role)
        if (userRoleRepository.existsByUsernameAndRole(username, role)) {
            throw DuplicatedEntityException("role for user")
        }
        userRoleRepository.save(entity)
    }

    @Transactional
    fun removeRoleFromUser(username: String, role: String) {
        if (!userRepository.existsByUsername(username)) {
            throw EntityNotFoundException("User $username")
        }
        if (!userRoleRepository.existsByUsernameAndRole(username, role)) {
            throw EntityNotFoundException("Role $role for user $username")
        }
        userRoleRepository.deleteByUsernameAndRole(username, role)
    }
}
